crates/ignore/src/types.rs

```diff
@@ -112,35 +114,241 @@ use crate::{Error, Match};
 /// The lifetime `'a` refers to the lifetime of the underlying file type
 /// definition, which corresponds to the lifetime of the file type matcher.
 #[derive(Clone, Debug)]
-pub struct Glob<'a>(GlobInner<'a>);
+pub struct Globs(GlobsInner);
 
 #[derive(Clone, Debug)]
-enum GlobInner<'a> {
+enum GlobsInner {
     /// No glob matched, but the file path should still be ignored.
     UnmatchedIgnore,
     /// A glob matched.
     Matched {
         /// The file type definition which provided the glob.
-        def: &'a FileTypeDef,
+        file_type_indices_token: FileTypeIndicesToken,
     },
 }
 
-impl<'a> Glob<'a> {
-    fn unmatched() -> Glob<'a> {
-        Glob(GlobInner::UnmatchedIgnore)
+impl Globs {
+    pub fn match_metadata_token(&self) -> Option<MatchMetadataToken> {
+        match &self.0 {
+            GlobsInner::Matched { file_type_indices_token } => {
+                Some(MatchMetadataToken::FileTypeIndicesToken(
+                    file_type_indices_token.clone(),
+                ))
+            }
+            _ => None,
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub enum MatchMetadataToken {
+    FileTypeIndicesToken(FileTypeIndicesToken),
+}
+
+const FILE_TYPE_INDICES_BIT_ARRAY_LEN: usize = 256;
```

Comment:
So the idea with the token avoiding allocation is that (at least in the case of multiple matching file-types) it would encode those matching indices in a fixed-length bit-vec (as long as those indices "fit" in that fixed-length, otherwise it falls back to allocating a `Vec<usize>`)

I came up with 256 here by looking at the length of the default file types that you add "all at once" when building a `WalkParallel` and rounding it up to the nearest power of 2

So since that is a known "fixed" length (that we always "use" in `tree-sitter-grep`) I think that guarantees that the found indices will always fit in a bit-vec of this size